qutip/core/data/convert.pyx

```diff
@@ -259,63 +261,171 @@ cdef class _to:
+                and return a data-layer type instance. The object may be
+                any object for which the condition function returned ``True``
+                when tested.
+
+            priority : positive real, optional (1)
+                The priority associated with this creator. Higher priority
+                conditions will be tested first and the first valid creator
+                (i.e. for which ``condition(object) == True``) will handle
+                the creation.
+
+        Notes
+        -----
+            Default creators are added with the following priorities:
+
+                * Objects that are instances of data-layer types are
+                  converted using ``.copy`` with priority ``100``.
+                * Objects that are ``numpy.ndarray`` instances or for
+                  which ``scipy.sparse.isspmatrix_csr`` is ``True`` are
+                  converted using an internal CSR converter with
+                  priority ``80``.
+                * Objects for which ``scipy.sparse.issparse`` is ``True``
+                  are converted using an internal CSR converter with
+                  priority ``10``.
+        """
+        _creators = []
+        for creator in creators:
+            if len(creator) == 2:
+                condition, create = creator
+                priority = 10
+            elif len(creator) == 3:
+                condition, create, priority = creator
+            if not callable(condition):
+                raise TypeError(str(condition) + " is not a callable")
+            if not callable(create):
+                raise TypeError(str(create) + " is not a callable")
+            _creators.append((condition, create, priority))
+        self._creators = sorted(_creators,
+                                key=lambda creator: creator[2],
+                                reverse=True)
+
+    def __call__(self, arg, shape=None):
+        for condition, create, _ in self._creators:
+            if condition(arg):
+                return create(arg, shape)
+        # Fallback on numpy if we don't have any matching creator method
+        return _data.Dense(np.array(arg, dtype=np.complex128),
+                           shape=shape, copy=False)
 
 to = _to()
 create = _create()
+
+import qutip.core.data as _data
+import numpy as np
+import scipy.sparse
+
+create.add_creators([
+    (
+        lambda arg: isinstance(arg, _data.base.Data),
+        lambda arg, shape: arg.copy(),
+        100
+    ),
+    (
+        lambda arg: scipy.sparse.isspmatrix_csr(arg),
+        lambda arg, shape: _data.CSR(arg, shape=shape),
+        80
+    ),
+    (
+        lambda arg: isinstance(arg, np.ndarray),
+        lambda arg, shape: _data.Dense(arg, shape=shape, copy=False),
+        80
+    ),
+    (
+        lambda arg: scipy.sparse.issparse(arg),
+        lambda arg, shape: _data.CSR(arg.tocsr(), shape=shape),
+        10
+    ),
+])
```

Comment:
Using lambdas here makes this object un-pickleable.  I think we should probably avoid that in the core library - I don't have a particular use-case in mind, though, so it's probably not necessary to change right now.